lib/pengine/unpack.c

```diff
@@ -573,17 +574,28 @@ unpack_nodes(xmlNode * xml_nodes, pe_working_set_t * data_set)
             new_node = NULL;
 
             id = crm_element_value(xml_obj, XML_ATTR_ID);
+            cluster_id = crm_element_value(xml_obj, PCMK_XA_CLUSTER_ID);
             uname = crm_element_value(xml_obj, XML_ATTR_UNAME);
             type = crm_element_value(xml_obj, XML_ATTR_TYPE);
             score = crm_element_value(xml_obj, XML_RULE_ATTR_SCORE);
-            crm_trace("Processing node %s/%s", uname, id);
+
+            crm_trace("Unpacking node %s with cluster ID %s and name %s",
+                      pcmk__s(id, "<none>"), pcmk__s(cluster_id, "same as id"),
+                      pcmk__s(uname, "same as id"));
 
             if (id == NULL) {
                 pcmk__config_err("Ignoring <" XML_CIB_TAG_NODE
                                  "> entry in configuration without id");
                 continue;
             }
-            new_node = pe_create_node(id, uname, type, score, data_set);
+
+            if (cluster_id == NULL) {
+                cluster_id = id;
+            }
+            if (uname == NULL) {
+                uname = id;
+            }
+            new_node = pe_create_node(cluster_id, uname, type, score, data_set);
```

Comment:
@kgaillot I suppose the new `cluster-id` attribute can supplant all internal uses of `id` for nodes? I hope so... there's nowhere to put a new attribute. `pe_node_s` and `pe_node_shared_s` are both public API and aren't created by constructors.

To confirm: After the change is complete, `id` is intended to be used only (1) for IDREFs and (2) as the default value for cluster-id and uname, correct? We should never need to use a node's `id` directly.

Reply:
I hope this doesn't get sticky when we create `<node>` elements in the CIB based on things like `crm_node_t:uuid` (e.g., `controld_join_dc.c:finalize_join_for()`). Lots to comb through.

Reply:
Gotcha.

Leaving this alone pending response to the second comment.

Reply:
> It's not guaranteed but should be unique in practice.

Unique among cluster layer IDs. As for unique among XML IDs... it should be in practice but it's just an integer (starting at 1). I don't like the idea of using that as an auto-generated XML ID that we'd then reference in other XML elements.

> However it can (at least theoretically) be NULL

Theoretically yes. In practice, when a `crm_node_t` gets allocated for a cluster node -- which happens only in `pcmk__get_peer()` or `known_node_cache_refresh_helper()` -- its `uuid` gets set to a string form of its Corosync node ID. That is, unless it's "not yet known by Corosync," which happens when `pcmk__get_peer()` is called with a 0 `id` argument. I'm not sure that can even happen while generating the `<nodes>` list.

If it *is* `NULL` (might not be possible), then we don't have a reasonable, non-arbitrary value to use as the XML ID. An arbitrary value isn't useful for IDREFs, since a cluster node should always be known by the cluster layer.

> I think if the node ID is XML ID conformant, we could set it as the XML ID and omit cluster-id, and otherwise add it as cluster-id and auto-generate an XML ID (probably as an actual UUID).

It's probably conformant based on the allowed characters, but it may not be conformant based on uniqueness. There may be another element somewhere in the XML with `id="<int>"`, however silly that might seem.
* Prevent the user from configuring any elements with integer IDs, in case a cluster layer ID conflicts with it? That's clunky and an arbitrary restriction to enforce.
* Allow the cluster layer to create an integer XML ID that turns out to be non-unique and causes a validation error? Bad.

> The node has to exist before it can be referenced, so users will be adding configuration references after the nodes have already been populated and the id is known.

Sure, but we can't use hashes because there could be a conflict. So we'd proably use something like `crm_generate_uuid()`, which means a node's XML ID would change each time it joined the cluster.

We could avoid that by never auto-generating a UUID, but that's problematic in other ways discussed above and below.

> The obj_ref "id" is constrained as ID type

text
```
            <element name="obj_ref">
              <attribute name="id"><text /></attribute>
            </element>
```

> (I didn't think about how this might impact `<node_state>`. There's no schema for the status section, so that's not a concern, but most likely the `<node_state>` XML ID has to match the XML ID, and that might open another can of worms.)

At least in `unpack_node_state()` (haven't checked exhaustively), we try looking up the node by `id` and then by `uname`. There is a fallback at least.

> I guess for Pacemaker Remote nodes, they will always need an auto-generated id and their name as cluster-id and uname.

Yep, and we'd need a way to make it persistent.

---

This is probably all doable if we keep the `<nodes>` list persistently regardless of current membership Keep the auto-generated XML ID the same until the node is removed by `crm_node -R` or some other tool, or until a node with a different uname joins with that cluster layer ID (in which case we generate a new UUID XML ID).

However we implement it, every IDREF to an auto-generated node XML ID would be a 36-character random string. It would technically be usable -- look up the XML ID of the node you're interested in, and put it there -- but obnoxious to visually parse. `pcs` and `crmsh` could maybe translate between node uname and node XML ID.

Reply:
Tags for nodes will probably look at least a little bit funny regardless... from skimming, it looks like right now if we're tagging a resource we'd have
```
<obj_ref id="my_rsc1"/>
```

while for a node we'd have the below, where you just have to know that "a bare integer ID usually means a node" without any further context clues from the ID
```
<obj_ref id="1"/>
```

Reply:
> Prevent the user from configuring any elements with integer IDs, in case a cluster layer ID conflicts with it? That's clunky and an arbitrary restriction to enforce.

Also, a future cluster layer may have IDs that aren't integers. Point is we don't know which cluster layer IDs will exist until all nodes join, and new ones may be added later.

Reply:
> > It's not guaranteed but should be unique in practice.
> 
> Unique among cluster layer IDs. As for unique among XML IDs... it should be in practice but it's just an integer (starting at 1). I don't like the idea of using that as an automatic XML ID that we'd then reference in other XML elements.

Yeah, cluster layer IDs could conflict with other non-node XML IDs. In this scenario though, we're auto-generating id as a true UUID, and using the cluster layer ID as cluster-id, so cluster-id should be unique. I'm not sure if there's a way to enforce an arbitrary attribute to be unique via the schema, but that would be a plus for the separate cluster-id idea.
 
> > However it can (at least theoretically) be NULL
> 
> Theoretically yes. In practice, when a `crm_node_t` gets allocated for a cluster node -- which happens only in `pcmk__get_peer()` or `known_node_cache_refresh_helper()` -- its `uuid` gets set to a string form of its Corosync node ID. That is, unless it's "not yet known by Corosync," which happens when `pcmk__get_peer()` is called with a 0 `id` argument. I'm not sure that can even happen while generating the `<nodes>` list.
> 
> If it _is_ `NULL` (might not be possible), then we don't have a reasonable, non-arbitrary value to use as the XML ID. We'd have to generate a temporary arbitrary value, which isn't useful for IDREFs since it should get overwritten when the node becomes known to the cluster layer.

We wouldn't generate temporary values, what we generated would become the permanent id. Once the cluster ID was known, it would be added as cluster-id.
 
> > I think if the node ID is XML ID conformant, we could set it as the XML ID and omit cluster-id, and otherwise add it as cluster-id and auto-generate an XML ID (probably as an actual UUID).
> 
> It's probably conformant based on the allowed characters, but it may not be conformant based on uniqueness. There may be another element somewhere in the XML with `id="<int>"`, however silly that might seem.

Yeah, we'd have to check for that. This is assuming we have the cluster ID when we create the node element; if we don't, we would use a true UUID.
 
>     * Prevent the user from configuring any elements with integer IDs, in case a cluster layer ID conflicts with it? That's clunky and an arbitrary restriction to enforce.

Definitely not. Any auto-generated XML ID should be a true UUID. (<node id> isn't currently an XML ID, so it doesn't matter for the current code.)
 
>     * Allow the cluster layer to create an integer XML ID that turns out to be non-unique and causes a validation error? Bad.

It would only be added as cluster-id
 
> > The node has to exist before it can be referenced, so users will be adding configuration references after the nodes have already been populated and the id is known.
> 
> Sure, but we can't use hashes because there could be a conflict. So we'd probably use something like `crm_generate_uuid()`, which means a node's XML ID would change each time it joined the cluster.

Nope, it's permanent. It's only auto-generated if it doesn't exist. Once it's generated, the CIB is saved and it never needs to be auto-generated again.
 
> > The obj_ref "id" is constrained as ID type
> 
> text
> 
> ```
>             <element name="obj_ref">
>               <attribute name="id"><text /></attribute>
>             </element>
> ```

Yep, that was my big oversight
 
> > (I didn't think about how this might impact `<node_state>`. There's no schema for the status section, so that's not a concern, but most likely the `<node_state>` XML ID has to match the XML ID, and that might open another can of worms.)
> 
> At least in `unpack_node_state()` (haven't checked exhaustively), we try looking up the node by `id` and then by `uname`. `id` would refer to the cluster layer ID, and there is a fallback to uname.

Yeah, how it should work under the new schema would be an open question
 
> > I guess for Pacemaker Remote nodes, they will always need an auto-generated id and their name as cluster-id and uname.
> 
> Yep, and we'd need a way to make it persistent.

It's called saving the CIB :)
 
> This is probably all doable if we keep the `<nodes>` list persistently regardless of current membership. Keep the auto-generated XML ID the same until the node is removed by `crm_node -R` or some other tool, or until a node with a different uname joins with that cluster layer ID (in which case we generate a new UUID XML ID).
> 
> However we implement it, every IDREF to an auto-generated node XML ID would be a 36-character random string. It would technically be usable -- look up the XML ID of the node you're interested in, and put it there -- but obnoxious to visually parse. `pcs` and `crmsh` could maybe translate between node uname and node XML ID.

Yeah it's probably not worth any effort

Reply:
> Yeah, cluster layer IDs could conflict with other non-node XML IDs. In this scenario though, we're auto-generating `id` as a true UUID, and using the cluster layer ID as `cluster-id`, so `cluster-id` should be unique. I'm not sure if there's a way to enforce an arbitrary attribute to be unique via the schema, but that would be a plus for the separate `cluster-id` idea.

`cluster-id` should be unique among `cluster-id` attributes, yes. Here, I was saying that if we DON'T want to auto-generate the XML ID as a UUID, then we'd have to use the cluster layer ID. The cluster layer ID is the only "real" ID that we might have. We don't configure `id` or `cluster-id` ourselves in the CIB -- `cluster-id` gets auto-populated from the cluster layer, and `id` either gets copied from that or auto-generated as a UUID.

I don't think this point really matters anymore, see below.

> We wouldn't generate temporary values, what we generated would become the permanent id. Once the cluster ID was known, it would be added as cluster-id.

Yeah I think I was wrong here. My uses of "temporary" and "permanent"/"persistent" are because I thought the existing `<nodes>` section might get overwritten or deleted-from, when certain events happen (when we populate the `<nodes>` section, when a node leaves the cluster, when Pacemaker restarts, etc.)

I only skimmed over these last night. I was thinking there are probably cases where we lose or overwrite a node from the `<nodes>` section and from the peer cache, and we'd need to generate a new XML ID when we create a new `<node>` element for it.

I don't clearly see anything like that at the moment. There are modify calls, but I don't see outright replacements, and the only removals I see are of unknown conflicting nodes.

> > There may be another element somewhere in the XML with id="<int>", however silly that might seem.

> Yeah, we'd have to check for that. This is assuming we have the cluster ID when we create the node element; if we don't, we would use a true UUID.

IMO it'd be cleaner to use a true UUID as `<node>` XML ID *unconditionally*. The alternative is to look for any other element in the CIB with an XML ID matching the `cluster-id` before using the `cluster-id` as the node's XML ID, and fall back to generating a UUID if there's a conflict... and then we may have some `<node>` elements with (for example) `id="1"` and other nodes with `id="3c7b763a-3..."` (36 chars).

That whole arrangement, with a possible mix of bare-int and UUID XML IDs, and with the need to unique-check the `cluster-id` first and try to use it, seems even more confusing than having every node use a UUID XML ID.

For slightly better readability, the XML ID could be `<uname>-<uuid>` if the uname exists and `<uuid>` otherwise.

---

Tags based purely on `uname` would be the most user-friendly way to tag nodes IMO. Obviously tags (as currently implemented) have to match against an `id` attribute, and `uname` isn't unique for use as an XML ID.

Maybe it would eventually be worth allowing tags based on some other attribute besides `id` (for example `uname`), with the caveat that each `obj_ref` may match more than one object. That could be a more convenient way to configure certain groupings for things like `crm_mon` output.

Reply:
> IMO it'd be cleaner to use a true UUID as `<node>` XML ID _unconditionally_.

agreed

fyi, long ago, it was common to use a true UUID or a large int derived from the IP address as node id (some regression tests still have them), so there is precedent. but i agree it would be a pain for users to copy and paste around all the time, not worth pursuing without demand